sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java

```diff
@@ -1402,6 +1407,15 @@ public void populateDisplayData(DisplayData.Builder builder) {
       return toBuilder().setConsumerFactoryFn(consumerFactoryFn).build();
     }
 
+    /**
+     * A custom {@link SerializableFunction} that determines whether the {@link ReadFromKafkaDoFn}
+     * should stop reading from the given {@link TopicPartition}.
+     */
+    public ReadSourceDescriptors<K, V> withCheckStopReadingFn(
```

Comment:
Yeah, will do. I'll also add support to commit transform together in this PR for reverting safety.

Reply:
I added more related javadoc around `KafkaIO`.

Thinking about the commit transform again. I don't think additional changes required here because:
* For stopped TopicPartition, we still want to commit the offset
* For deleted TopicPartition, we should get exceptions from Kafka Consumer API, which has been handled in the Commit transform.

I'll merge this PR if all these look good to you : )